DO NOT publish openapi specs containing bad types

yue9944882 opened the pull request:
for now, we can set arbitrary types for the json schema like:

```yaml
  validation:
    openAPIV3Schema:
      properties:
        spec:
          type: bug // whatever type except for the root element
      type: object
```
this is violating https://tools.ietf.org/html/draft-wright-json-schema-00#section-4.2 and this pull adds a list of supported types and their validation according to the specification.

https://swagger.io/docs/specification/data-models/data-types/

```release-note
NONE
```

/king bug
/sig api-machinery

/cc @sttts

yue9944882 commented:
/assign @sttts

yue9944882 commented:
/retest

yue9944882 commented:
/test pull-kubernetes-bazel-build

yue9944882 commented:
/kind bug
/area custom-resources

yliaog commented:
/cc @liggitt @roycaihw

liggitt commented:
/hold

this is a breaking change, so it cannot go in unconditionally.

what is the behavior:
1. of the custom resource validation?
2. of the openapi publishing?

we should ensure:
1. v1 CRDs do not permit introduction of invalid types
2. a CRD previously persisted with an invalid type does not crash the server when validating custom resources

roycaihw commented:
1. of the custom resource validation?
- CR gets created successfully if the field is optional and the CR blob omits that field
- otherwise [go-openapi](https://github.com/kubernetes/kubernetes/blob/a5e3d7469149c704e32a9fd858be2a25c4ca9a2c/vendor/github.com/go-openapi/validate/type.go#L174-L176) infers schema type from input CR blob and fails validating the type
```
validation failure list:
spec in body must be of type bug // whatever type except for the root element: "object"
```
2. of the openapi publishing?
[kubectl](https://github.com/kubernetes/kubernetes/blob/38e8b8a9f65e66b779a0985e17868b036c1d601b/vendor/k8s.io/kube-openapi/pkg/util/proto/document.go#L161) fails parsing the non-standard type
```
Unknown primitive type: "bug // whatever type except for the root element"
```

yue9944882 commented:
yea.. we have to ensure backward-compatibility in this fix. from a slack discuss w/ @sttts, we can only forbid invalid-typed CRD under following cases, so that existing CRD will keep working:

- creates new CRD w/ invalid-typed property
- updates valid-typed property on an existing CRD to invalid

yue9944882 commented:
> of the openapi publishing?
kubectl fails parsing the non-standard type
Unknown primitive type: "bug // whatever type except for the root element"

confirmed this one, kubectl aborts after requesting `/openapi/v2` leaving an schema error in this case:

> error: SchemaError(example.v1alpha1.Cluster.spec): Unknown primitive type: "bug"

liggitt commented:
> confirmed this one, kubectl aborts after requesting `/openapi/v2` leaving an schema error in this case:
> 
> > error: SchemaError(example.v1alpha1.Cluster.spec): Unknown primitive type: "bug"

we are free to refuse to publish openapi for a CRD if it has an invalid type

yue9944882 commented:
added backward-compatibility and couple of test, a polite ping for review

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go:
```diff
@@ -366,7 +371,26 @@ func ValidateCustomResourceDefinitionSpecUpdate(spec, oldSpec *apiextensions.Cus
+			for j, oldCRDVersion := range oldSpec.Versions {
+				if mungedCRDVersion.Name == oldCRDVersion.Name {
+					if mungedCRDVersion.Schema != nil && oldCRDVersion.Schema != nil {
+						dropInvalidJSONSchemaFields(mungedCRDSpec.Versions[i].Schema.OpenAPIV3Schema, oldSpec.Versions[j].Schema.OpenAPIV3Schema)
```
mutating a copy of the spec inside validation is not what I expected. the changes I expected:

1. do not publish openapi for a schema containing bad types
2. do not change validation (yet)
3. in v1 validation only (https://github.com/kubernetes/kubernetes/pull/79604) require valid object types on create, and do not allow changing valid to invalid on update

roycaihw replied:
for publishing openapi we could do something like https://github.com/roycaihw/kubernetes/commit/2a75b8e8fabc8eeaf72b6808714fb536e81a3269. It's probably better to do the validation on JSONSchemaProps (like what you have here) instead of structural schema so we can save some code between schema validation and publishing

yue9944882 replied:
>  It's probably better to do the validation on JSONSchemaProps (like what you have here) instead of structural schema

validating structural really saves code comparing to JSONSchemaProps at least at this point, b/c the json model is far more complex and most fields are even unused/unsupported actually. i just tried another approach basing on your commit.

roycaihw replied:
I had `old *Structural` in the interface because I had [2 & 3](https://github.com/kubernetes/kubernetes/pull/79587#discussion_r300005679) in my mind. For openapi publishing old is always nil (treated as CREATE)

yue9944882 replied:
> I had 2 & 3 in my mind. For openapi publishing old is always nil (treated as CREATE)

oh thanks i mis-understood. if we will apply the validation in v1, it should better be validating JSONSchemaProps, isn't it?

roycaihw replied:
yes. JSONSchemaProps needs a backwards-compatible validation in v1; in v1beta1, the openapi controller can simply ignore invalid schemas by validating either JSONSchemaProps or Structural type

roycaihw committed:
do not publish openapi for a schema containing bad types

minor rewords & refactor

run script: update misc

remove null from supported types

yue9944882 commented:
/test pull-kubernetes-integration

yue9944882 commented:
/retest

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder.go:
```diff
@@ -59,17 +59,17 @@ var buildDefinitions sync.Once
 
 // BuildSwagger builds swagger for the given crd in the given version
-func BuildSwagger(crd *apiextensions.CustomResourceDefinition, version string) (*spec.Swagger, error) {
+func BuildSwagger(crd *apiextensions.CustomResourceDefinition, oldStructuralSchema *structuralschema.Structural, version string) (*spec.Swagger, *structuralschema.Structural, error) {
```
needing to pass an old schema into this method doesn't make sense to me. if there are invalid types in the schema, we should not publish openapi for it

roycaihw replied:
yes for openapi publishing we can stop publishing invalid schema and it's [backwards compatible](https://github.com/kubernetes/kubernetes/pull/79587#issuecomment-507462802)

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go:
```diff
@@ -184,18 +190,25 @@ func (c *Controller) sync(name string) error {
 		}
-		spec, err := BuildSwagger(crd, v.Name)
+		var oldStructuralSchema *structralschema.Structural
+		if oldSpecs[v.Name] != nil {
```
and maintaining old schemas here doesn't make sense to me either

yue9944882 commented:
/retest

yue9944882 commented:
@liggitt @roycaihw i picked the commit and appended some minor changes

roycaihw commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/validation.go:
```diff
@@ -75,6 +76,46 @@ func ValidateStructural(s *Structural, fldPath *field.Path) field.ErrorList {
+
+	allErrs := field.ErrorList{}
+
+	if old != nil {
```
Ah, if the scope of this pr is only in openapi publishing, we can drop the `old *Structural` bit to make it more concrete and less confusing. Given that traversing JSONSchemaProps seems to more complex than traversing Structural schema, I'm not sure how re-usable this code will be in v1. Otherwise LGTM

I'll let @liggitt to make the final call

liggitt replied:
let's limit this to publishing. I'll tighten validation for v1 CRDs in https://github.com/kubernetes/kubernetes/pull/79604

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/validation.go:
```diff
@@ -75,6 +76,46 @@ func ValidateStructural(s *Structural, fldPath *field.Path) field.ErrorList {
+}
+
+var jsonSchemaSupportedTypes = sets.NewString(
+	"null", "boolean", "integer", "object", "array", "number", "string",
```
null is not a valid openapi v3 type

yue9944882 replied:
removed "null"

yue9944882 commented:
politely bumping

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/validation.go:
```diff
@@ -75,6 +76,46 @@ func ValidateStructural(s *Structural, fldPath *field.Path) field.ErrorList {
+// ValidateStructuralUpdate checks invariants of a structural schema on update or creation.
+// in publishing, we make old being nil to keep backwards compatibility.
+// TODO: reuse the function in v1 validation
+func ValidateStructuralUpdate(s *Structural, old *Structural, fldPath *field.Path) field.ErrorList {
```
I don't think this needs both the old and new object. Something simpler like checking if the schema has invalid types and avoiding serving openapi could be done with a method like this (included in https://github.com/kubernetes/kubernetes/pull/79604):

https://github.com/liggitt/kubernetes/commit/aeeaed6a7f750acd3afb900c35550ff4997ded30#diff-54fdd67617b0f2ef1b5a09148f77303eR1198

yue9944882 replied:
are you suggesting that this pull should be rebasing after #79604?

liggitt replied:
This PR can proceed first. It's fine for it to lift that method and export it (or place it in a package where that validation method could access it).

yue9944882 replied:
copied and exported from validation pkg

liggitt commented:
/hold cancel
/lgtm
/approve

liggitt commented:
/priority important-soon
